### PR TITLE
Begin adding the `OnClick` actions

### LINKED ISSUE #81: Seperate functions for each part of the transform in `SetTransform`
`SetTransform` function can be left alone, "inner" functions added for each piece of functionality. This way the separate functions can be called individually as well.

- [ ] Parent
- [ ] Scale
- [ ] Position
- [ ] Rotation

The inner functions should return their values whereas `SetTransform` is void. This will help with instantiating a prefab (canvas, text, etc)

### LINKED ISSUE #82: Disable button after first click if `!RemainEnabled`
I believe I'll have to write a function to disable the button

### LINKED ISSUE #83: Add `<Link>` Actions
Each action is a type of `Actions` class. Will have to figure out the `Clicks` property.

Each action get's added to the `OnClick` property of the `Button`

### PR DESCRIPTION
**Pull request by RobertGemmaJr**
- Adds the skeleton code for the link actions (`SetLinkActions`)
- Adds `ActionMethods.cs`
- `gameObjects` dictionary is now a `<string, (GameObject, Object)>` pair
- Updates `unity.md`

I've realized in working with this that I need to generate all of the GameObjects/Sounds/Timelines/etc because the actions reference them. I decided it's best to cut this PR a little short and do some refactoring in different branches before finishing off the Action types in `sample.xml`

closes #81
closes #82 
closes #83

### CODE REVIEW

File: unity/CAVE/Assets/W3D_Translator/CLI.cs [hunk 1]
```diff
@@ -200,11 +214,9 @@ private static void BuildWalls(Story xml, Transform storyT)
     }
 
     // Convert Story.ObjectRoot to a dictionary of {name: GameObject} pairs
-    private static Dictionary<string, GameObject> TranslateGameObjects(
-        List<W3D.Object> objectList, GameObject story
-    )
+    private static ObjDictionary TranslateGameObjects(List<W3D.Object> objectList, GameObject story)
     {
-        Dictionary<string, GameObject> gameObjects = new();
+        ObjDictionary gameObjects = new();
```
Comment (broarr) [inline unity/CAVE/Assets/W3D_Translator/CLI.cs:219]: Why do you need this special class if there's no special methods for the new class?

File: unity/CAVE/Assets/W3D_Translator/CLI.cs [hunk 2]
```diff
@@ -241,25 +251,76 @@ private static void BuildWalls(Story xml, Transform storyT)
...
 
+    private static void SetLinkActions(ObjDictionary gameObjects, GameObject story)
+    {
+        ActionMethods methods = story.GetComponent<ActionMethods>();
+
+        foreach (KeyValuePair<string, (GameObject, W3D.Object)> pair in
+            gameObjects.Where(pair => pair.Value.Item2.LinkRoot is not null)
+        )
+        {
+            (GameObject go, W3D.Object obj) = pair.Value;
```
Comment (broarr) [inline unity/CAVE/Assets/W3D_Translator/CLI.cs:273]: Can you do `var (go, obj) = pair.Value` here? Or do you need to specify types? In general you're specifying the type a lot when you can have C# do that work for you

File: unity/CAVE/Assets/W3D_Translator/CLI.cs [hunk 3]
```diff
@@ -241,25 +251,76 @@ private static void BuildWalls(Story xml, Transform storyT)
...
+                        // TODO 91
+                        break;
+                    default:
+                        if (xmlAction.Type == Actions.ActionTypes.Restart)
+                        {
+                            // TODO 92
+                        }
+                        else { throw new Exception("TODO"); }
+                        break;
+                }
```
Comment (broarr) [inline unity/CAVE/Assets/W3D_Translator/CLI.cs:318]: Should the actions be within the if statement above? We only want an action to happen if the right number of clicks happens right?

**Review by broarr — APPROVED** [review]
This looks great, I just had one clarifying question about the button action logic

File: unity/CAVE/Assets/W3D_Translator/CLI.cs (hunk 2 above)
Comment (RobertGemmaJr) [reply unity/CAVE/Assets/W3D_Translator/CLI.cs:273]: I can but get a linter warning to specify the type when I do that. Do you think it's safe to disable the warning? I don't really mind all of the type definitions

File: unity/CAVE/Assets/W3D_Translator/CLI.cs (hunk 1 above)
Comment (RobertGemmaJr) [reply unity/CAVE/Assets/W3D_Translator/CLI.cs:219]: Looks like I can use a `using` directive? https://stackoverflow.com/questions/12099271/simplify-building-dictionaries

Comment (broarr) [reply unity/CAVE/Assets/W3D_Translator/CLI.cs:219]: That's why I like type inference so much! 

```
var gameObjects = new Dictionary<string, GameObject>();
```

You've got the ugly definition once, but the compiler tracks it for the rest of the life of that variable

Comment (broarr) [reply unity/CAVE/Assets/W3D_Translator/CLI.cs:219]: > Looks like I can use a `using` directive? https://stackoverflow.com/questions/12099271/simplify-building-dictionaries

Interesting! I've never seen `using` used like that!

File: unity/CAVE/Assets/W3D_Translator/CLI.cs (hunk 2 above)
Comment (broarr) [reply unity/CAVE/Assets/W3D_Translator/CLI.cs:273]: booo, I don't think these types are helpful, and the compiler can figure them out. I'd personally remove that rule, but it's up to you! If you like type definitions you can keep them

File: unity/CAVE/Assets/W3D_Translator/CLI.cs (hunk 3 above)
Comment (RobertGemmaJr) [reply unity/CAVE/Assets/W3D_Translator/CLI.cs:318]: I'm 99% sure no, although I'm not entirely sure how the number of clicks is going to play out.

Here we're just adding callback functions to the buttons onClick, similar to web dev. The difference being that in Unity `onClick` is a list - you can add as many callback functions as you'd like. The actual code looks like 

```c#
            UnityEventTools.AddObjectPersistentListener(
                onClick, // list of callback functions
                new UnityAction<Button>(methods.DisableButton), // Callback function to add
                button // Parameter of methods.DisableButton
            );
```

The question remains as to what will go in the `if` block - how we'll tell Unity to fire the onClick. It's looking a little like I'd add a counter function to the onClick itself and manage the callback functions in a different list? Or I'd capture `onClick.Invoke()` and manage that based on the counter? Not sure how they'll play out but the list of callback functions and how/when they're execute definitely seem to be different tasks

Comment (RobertGemmaJr) [reply unity/CAVE/Assets/W3D_Translator/CLI.cs:318]: https://gamedev.stackexchange.com/questions/163317/trigger-the-buttons-onclick-after-checking-that-the-cursor-has-been-on-it-for-s

https://stackoverflow.com/questions/56323774/unity-how-can-i-make-a-function-fire-only-when-player-clicked-on-it-3-times

File: unity/CAVE/Assets/W3D_Translator/CLI.cs (hunk 2 above)
Comment (RobertGemmaJr) [reply unity/CAVE/Assets/W3D_Translator/CLI.cs]: Was pretty easy to disable in the `editorconfig`!

**Pull request merged**